py/server/deephaven/table.py

```diff
@@ -503,6 +505,112 @@ def meta_table(self) -> Table:
+        the shared lock of the update graph before reading the table data. This provides a consistent view of the data.
+        The side effect of this is that the table will not be able to refresh while the table is being iterated on.
+        Additionally, the generator internally maintains a fill context. The auto acquired shared lock and the fill
+        context will be released after the generator is destroyed. That can happen implicitly when the generator
+        is used in a for-loop. When the generator is not used in a for-loop, to prevent resource leaks, it must be closed
+        after use by either (1) by setting it to None, (2) using the del statement, or (3) calling the close() method on it.
+
+        Args:
+            cols (Optional[Union[str, Sequence[str]]]): The columns to read. If None, all columns are read.
+
+        Returns:
+            A generator that yields a dictionary of column names to scalar values.
+        """
+        from deephaven._table_reader import _table_reader_rows # to prevent circular import
+
+        return _table_reader_rows(self, cols)
+
+    def iter_tuple(self, cols: Optional[Union[str, Sequence[str]]] = None, name: str = 'Deephaven') -> Generator[Tuple[Any], None, None]:
+        """ Returns a generator that reads one row at a time from the table into a named tuple. The named tuple is made
+        up of fields with their names being the column names and their values being of the column data types.
+
+        If the table is refreshing and no update graph locks are currently being held, the generator will try to acquire
+        the shared lock of the update graph before reading the table data. This provides a consistent view of the data.
+        The side effect of this is that the table will not be able to refresh while the table is being iterated on.
+        Additionally, the generator internally maintains a fill context. The auto acquired shared lock and the fill
+        context will be released after the generator is destroyed. That can happen implicitly when the generator
+        is used in a for-loop. When the generator is not used in a for-loop, to prevent resource leaks, it must be closed
+        after use by either (1) by setting it to None, (2) using the del statement, or (3) calling the close() method on it.
+
+        Args:
+            cols (Optional[Union[str, Sequence[str]]]): The columns to read. If None, all columns are read. Default is None.
+            name (str): The name of the named tuple. Default is 'Deephaven'.
+
+        Returns:
+            A generator that yields a named tuple or regular tuple for each row in the table
+        """
+        from deephaven._table_reader import _table_reader_rows # to prevent circular import
+        named_tuple_class = namedtuple(name, cols or [col.name for col in self.columns], rename=False)
+
+        for row in _table_reader_rows(self, cols):
+            yield named_tuple_class(**row)
```

Comment:
I actually thought about it but we need an intermediate data structure to hold the values anyways before packing them into a tuple. Although we could use a List instead of Dict to save a bit space etc., the trade-off of just using dict and having a simplified code is better IMO.

Reply:
I don't have a great solution to easily avoid the garbage.  My best ideas involve using `exec` plus code string generation.  I would think about this path for a few minutes to see if  it happens to be easy and good before deciding to leave what we have.